sky/clouds/gcp.py

```diff
@@ -73,6 +73,7 @@
     ~/google-cloud-sdk/install.sh -q >> {_GCLOUD_INSTALLATION_LOG} 2>&1 && \
     echo "source ~/google-cloud-sdk/path.bash.inc > /dev/null 2>&1" >> ~/.bashrc && \
     source ~/google-cloud-sdk/path.bash.inc >> {_GCLOUD_INSTALLATION_LOG} 2>&1; }}; }} && \
+    export GOOGLE_APPLICATION_CREDENTIALS=~/.config/gcloud/application_default_credentials.json && \
```

Comment:
This can fix credential issue for #4512 cc @weih1121 

failure tests: `pytest tests/smoke_tests/test_mount_and_storage.py::test_azure_storage_mounts_with_stop --azure`

Reply:
Have we tested this without using a service account, e.g., with a normal GCP authentication locally or with `gcp.remote_identity: SERVICE_ACCOUNT`?

We should not assume that this file exists on the remote.

Reply:
1) If the file ~/.config/gcloud/application_default_credentials.json does not exist, or it exist but invalud, the sky launch will fail because GCP is not configured.

```+ [ ! -e ~/tmp-workdir/circle-link ] && ln -s ~/tmp-workdir/ ~/tmp-workdir/circle-link || true
+ touch ~/.ssh/id_rsa.pub
+ sky launch -y -c t-azure-storage-mount-b8-31 --cloud azure /var/folders/83/zxqx914s57x310rfnhq8kk9r0000gn/T/tmpy0nu9vj0.yaml
D 12-31 14:55:23 skypilot_config.py:228] Using config path: /Users/zepingguo/.sky/config.yaml
D 12-31 14:55:23 skypilot_config.py:233] Config loaded:
D 12-31 14:55:23 skypilot_config.py:233] {}
D 12-31 14:55:23 skypilot_config.py:245] Config syntax check passed.
Task from YAML spec: /var/folders/83/zxqx914s57x310rfnhq8kk9r0000gn/T/tmpy0nu9vj0.yaml
WARNING:google.auth.compute_engine._metadata:Compute Engine Metadata server unavailable on attempt 1 of 3. Reason: timed out
WARNING:google.auth.compute_engine._metadata:Compute Engine Metadata server unavailable on attempt 2 of 3. Reason: timed out
WARNING:google.auth.compute_engine._metadata:Compute Engine Metadata server unavailable on attempt 3 of 3. Reason: timed out
WARNING:google.auth._default:Authentication failed using Compute Engine authentication due to unavailable metadata server.
WARNING:google.auth.compute_engine._metadata:Compute Engine Metadata server unavailable on attempt 1 of 3. Reason: timed out
WARNING:google.auth.compute_engine._metadata:Compute Engine Metadata server unavailable on attempt 2 of 3. Reason: timed out
WARNING:google.auth.compute_engine._metadata:Compute Engine Metadata server unavailable on attempt 3 of 3. Reason: timed out
WARNING:google.auth._default:Authentication failed using Compute Engine authentication due to unavailable metadata server.
Traceback (most recent call last):
  File "/Users/zepingguo/miniconda3/envs/sky/bin/sky", line 8, in <module>
    sys.exit(cli())
  File "/Users/zepingguo/.local/lib/python3.10/site-packages/click/core.py", line 1157, in __call__
    return self.main(*args, **kwargs)
  File "/Users/zepingguo/.local/lib/python3.10/site-packages/click/core.py", line 1078, in main
    rv = self.invoke(ctx)
  File "/Users/zepingguo/Desktop/skypilot/sky/utils/common_utils.py", line 366, in _record
    return f(*args, **kwargs)
  File "/Users/zepingguo/Desktop/skypilot/sky/cli.py", line 838, in invoke
    return super().invoke(ctx)
  File "/Users/zepingguo/.local/lib/python3.10/site-packages/click/core.py", line 1688, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/Users/zepingguo/.local/lib/python3.10/site-packages/click/core.py", line 1434, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/Users/zepingguo/.local/lib/python3.10/site-packages/click/core.py", line 783, in invoke
    return __callback(*args, **kwargs)
  File "/Users/zepingguo/Desktop/skypilot/sky/utils/common_utils.py", line 386, in _record
    return f(*args, **kwargs)
  File "/Users/zepingguo/Desktop/skypilot/sky/cli.py", line 1118, in launch
    task_or_dag = _make_task_or_dag_from_entrypoint_with_overrides(
  File "/Users/zepingguo/Desktop/skypilot/sky/cli.py", line 790, in _make_task_or_dag_from_entrypoint_with_overrides
    dag = dag_utils.load_chain_dag_from_yaml(entrypoint, env_overrides=env)
  File "/Users/zepingguo/Desktop/skypilot/sky/utils/dag_utils.py", line 101, in load_chain_dag_from_yaml
    task = task_lib.Task.from_yaml_config(task_config, env_overrides)
  File "/Users/zepingguo/Desktop/skypilot/sky/task.py", line 438, in from_yaml_config
    storage_obj = storage_lib.Storage.from_yaml_config(storage[1])
  File "/Users/zepingguo/Desktop/skypilot/sky/data/storage.py", line 1045, in from_yaml_config
    storage_obj = cls(name=name,
  File "/Users/zepingguo/Desktop/skypilot/sky/data/storage.py", line 560, in __init__
    self.add_store(StoreType.GCS)
  File "/Users/zepingguo/Desktop/skypilot/sky/data/storage.py", line 896, in add_store
    store = store_cls(
  File "/Users/zepingguo/Desktop/skypilot/sky/data/storage.py", line 1551, in __init__
    super().__init__(name, source, region, is_sky_managed,
  File "/Users/zepingguo/Desktop/skypilot/sky/data/storage.py", line 263, in __init__
    self._validate()
  File "/Users/zepingguo/Desktop/skypilot/sky/data/storage.py", line 1596, in _validate
    raise exceptions.ResourcesUnavailableError(
sky.exceptions.ResourcesUnavailableError: Storage 'store: gcs' specified, but GCP access is disabled.```

2) If file `~/.config/gcloud/application_default_credentials.json` is configured as personal account and valid, it will be successful

```
(sky) ➜  skypilot git:(dev/zeping/fix_smoke_tests) pytest tests/smoke_tests/test_mount_and_storage.py::test_azure_storage_mounts_with_stop --azure
D 12-31 15:02:46 skypilot_config.py:228] Using config path: /Users/zepingguo/.sky/config.yaml
D 12-31 15:02:46 skypilot_config.py:233] Config loaded:
D 12-31 15:02:46 skypilot_config.py:233] {}
D 12-31 15:02:46 skypilot_config.py:245] Config syntax check passed.
bringing up nodes...
[azure_storage_mounts] Test started. Log: less /var/folders/83/zxqx914s57x310rfnhq8kk9r0000gn/T/azure_storage_mounts-s5h1dawv.log
[azure_storage_mounts] Passed.
[azure_storage_mounts] Log: less /var/folders/83/zxqx914s57x310rfnhq8kk9r0000gn/T/azure_storage_mounts-s5h1dawv.log
[azure_storage_mounts] 
.
1 passed, 272 warnings in 468.11s (0:07:48)
(sky) ➜  skypilot git:(dev/zeping/fix_smoke_tests)
```

3) If file `~/.config/gcloud/application_default_credentials.json` is configured as service account and valid, it will also be successful


We don't assume this file exists. If the file does not exist, the credential may fail as usual. If the file exists, the credential may succeed.